### ISSUE TITLE
Infinite file uploads

### ISSUE THREAD

**Issue by lastmjs** [demergent-labs/azle#1660]
We can already upload files on the initial deploy, up to 90 MiB during the Rust compilation to Wasm. These assets are specified with the assets property of dfx.json. We would like to allow larger files to be uploaded. To do this, I am thinking that they must be chunk uploaded in chunks of 2 MiB to stay under the message limit, and they must be aggregated in the canister and written to the file system.

Keep in mind authorization, we will need to figure this out. The Azle autoreload functionality is doing chunk uploading of just one file, and it doesn't write that file to the file system in the canister, it just keeps it in memory and then uses that to reload the JS runtime. It uses an agent locally in the Nodejs process that does the uploading, but it does so without any kind of authentication.

For these large files to work we will need authentication and to write the files to the filesystem. It would also be great to have these files, or maybe the assets on deploy, automatically uploaded somehow if they are changed...hmmm...maybe we should think about that one. If the files are very large then perhaps we would not want them to be automatically uploaded.

Anyhow, thinking about automatic uploading of assets is probably part of this issue, no assets are currently uploaded during auto-reload even if they are small and quick. This could be desirable for example for a frontend using Express static.

- [x] src/compiler/file_watcher/file_watcher.ts shows how to do a basic chunk upload
- [x] Write everything into the filesystem in the canister
- [x] Basic unauthenticated `assets_large` dfx.json API
- [x] Create post_install node script
- [x] Upload entire directories
- [x] clear the global hashmaps at the end and/or as we go
- [x] Create npx command that will automatically upload files into an azle canister, just point and shoot
- [x] chunk write the files to disk
- [x] Add throttling so we don't overload the replica as per [this forum post](https://forum.dfinity.org/t/cant-deploy-large-wasm-binary-429-too-many-requests/27721/4?u=lastmjs)
- [x] improve signature of post install script
     - [x] combine the two scripts
- [x] make sure all our reads and writes are streams and can handle > 2GiB files
- [x] push it to it's max (test until it breaks, if 1GB works, to a 2 etc until it breaks)
- [x] Be able to upload any size file up to the heap limit (~4 GiB)
- [x] stable storage
     - [x] hook up fs to stable storage
     - [x] write temp chunks to fs
     - [x] write hashes to stable memory
     - [x] check hashes before uploading
     - [x] #1972
     - [x] update wsam-forge/stable-fs to version that has fix
          - [x] figure out bug where 100KiB file is not uploaded properly
          - [x] figure out why create_dir_all is failing sometimes with a "already exists" error
- [x] be able to tests await in order, vs set time out, vs a manual call that returns if you need to call again
- [x] make a hash of the file as you go for testing
- [x] see if there is a way to store that hash for later if we want to compare hashes to determine if the file needs to be uploaded or not.
- [x] find all the possible bottle necks when large files come into play
- [x] Tackle authentication
- [x] #1968
- [x] universal authentication
- [x] hashing process checking at the end seems a bit complex maybe slow
- [x] Add File hashing finished log at the end
- [x] #1969
- [x] rename public methods, prefix with `_azle` to not have developer clashes
- [x] Consider making tests more robust for stable memory filesystem once we enable it, because the dfx --upgrade-unchanged tests will probably pass no matter what, which could be brittle in actually testing that the functionality is present and working
- [x] We should probably rename the is_authenticated guard function to something like `controller_guard` or `guard_against_non_controllers`
- [x] Please make sure that in the macro attribute hidden is true in all of the canister methods that have been added for file uploading
- [x] We should not write Rust code inside of quote if we don't need to do it, I think a lot of the Rust code was written inside of quotes by mistake
- [x] Use --output json for any dfx canister calls so that we can avoid strange parsing of the responses

**Comment by bdemann** [comment]
### Time trial info
- A file with 100MB of random bytes was broken up into 53 chunks and was uploaded successfully
- A file with 175MB of random bytes was broken up into 92 chunks and was not
    - Of the upload updates called only one had to be retried from a 429 (Too Many Requests) error
    - All of the chunks uploaded fine but the write broken when it tired to collect the various parts into one list

**Comment by bdemann** [comment]
In the current system it takes about 15 seconds to upload and write a 152 MB file. Adding a single await took our time to upload the same file to 2 minutes and 50 seconds. I'm going to go ahead and work towards chunk writing the files

**Comment by bdemann** [comment]
With throttling a 175 MiB file took less than 56 seconds to upload
With throttling a 152 MiB file took 48 seconds to upload
Without throttling (because the await is enough throttling on it's own) but with chunked uploading the 152 MiB upload awaited was 2 minutes 54 second.

**Comment by bdemann** [comment]
Time trials:
|Size| Background w/ throttling | Await throttled by await |
| ---- | --------------------------------------| -----------------------------------|
| 1 GB | 5 min | 20 min|
| 100 MB | 30 sec| 2 min|
| 10 MB | 4 sec | 12 sec |
| 1 MB | 50 ms |60 ms |

**Comment by bdemann** [comment]
Important! I don't think we can hash as we go, because the order we do the hash in matters. So just like writing, we need to collect the hashes and do them all afterwards at the end.

**Comment by bdemann** [comment]
I had this idea while we for tests back before we had to really do a lot of throttling. The idea was to check if the file existed before proceeding with the test, and if not, wait a little bit before failing the test, just in case it was still in the middle of an upload. One limitation is that since we have to write in chunks the file will exist before the file is completely written to. So maybe we could do something to work around that. Either way, with the trotting as it now is, it doesn't matter, waiting for a couple of seconds at the start is good enough.

```javascript
  let attempts = 0;
  while (exits === false) {
      await new Promise((resolve) => setTimeout(resolve, 5000));
      let response = await fetch(
          `${origin}/exists?path=${canisterFilePath}`
      );
      exits = await response.json();
      if (attempts++ > 2) {
          return {
              Err: `File ${canisterFilePath} failed to upload`
          };
      }
  }
```

**Comment by bdemann** [comment]
One of the things @lastmjs wants to do is write the file to disk as we go instead 
of writing all the chunks after we receive them. The idea was to keep a list of all 
of the chunks we received and what size they were so as we got another chunk, 
we could write that chunk to the appropriate place in the file.

As far as I can tell there is no way to just insert the contents into the middle
of a file. Every technique I have found for doing this requires some sort of
shifting or copying the existing values to later on in the file to do the insertion. 
So in the event that need to write a new chunk into the middle of a large file 
we would quickly run out of cycles as we try and read the rest of the file into 
memory, or if not when reading it into memory, then surely as we were 
writing it back to disk.

So if we want to write the file as we go then we have to make sure the chunks
come to us in the right order, or we need to stop and they need to try again.

With the throttling we are really very likely to receive the chunks in order.
But if somehow one sneaks out of order, that would cause a lot of possible
retries on the client end. If it throws off their timestamps then it will be
a complete restart. But it could be a requirement we make for ourselves.

Awaiting each call to the canister would insure that we are sending them in the
right order, but that would drastically slow down the process.
